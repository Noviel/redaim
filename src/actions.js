import { TRIGGER_UNKNOWN, TARGET_UNKNOWN, EXEC_UNKNOWN } from './constants';

export const createAction = 
  (trigger = TRIGGER_UNKNOWN) => 
  (target = TARGET_UNKNOWN) => 
  ({ exec = EXEC_UNKNOWN, payload = null, meta, ...rest } = {}) => 
({
  ...rest,
  type: trigger,
  payload,
  meta: {
    ...meta,
    target: target,
    exec
  }
});

export const aimFromAction = ({ type = '', payload = null, meta, ...rest } = {}) => {
  const [ 
    trigger = TRIGGER_UNKNOWN, 
    target = TARGET_UNKNOWN, 
    exec = EXEC_UNKNOWN 
  ] = type.split('/');

  return {
    ...rest,
    type: trigger || TRIGGER_UNKNOWN,
    payload,
    meta: {
      ...meta,
      target: target || TARGET_UNKNOWN,
      exec: exec || EXEC_UNKNOWN
    }
  };
};

export const wrapActions = fn => actions => {
  const finalActions = {};
  
  for (const a in actions) {
    const currAction = actions[a];
    if (typeof currAction === 'function') {
      finalActions[a] = (...args) => fn(currAction(...args));
    } else {
      finalActions[a] = (...args) => fn(currAction, ...args);
    }
  }

  return finalActions;
};

export const bindActions = actionCreatorWithTrigger => target => actions =>
  wrapActions(actionCreatorWithTrigger(target))(actions);
  
export const createActions = trigger => {
  const boundToTrigger = bindActions(createAction(trigger));

  return target => { 
    const boundToTarget = boundToTrigger(target);

    return actions => boundToTarget(actions);
  };
};

// reversed order of arguments
  
export const createActionR =
  ({ exec = EXEC_UNKNOWN, payload, meta, ...rest } = {}) =>
  (trigger = TRIGGER_UNKNOWN) =>
  (target = TARGET_UNKNOWN) =>
({
  ...rest,
  type: trigger,
  payload,
  meta: {
    ...meta,
    target: target,
    exec
  }
});

export const wrapActionsR = actions => fn => {
  const finalActions = {};
  
  for (const a in actions) {
    const currAction = actions[a];
    if (typeof currAction === 'function') {
      finalActions[a] = (...args) => fn(currAction(...args));
    } else {
      finalActions[a] = (...args) => fn(currAction, ...args);
    }
  }

  return finalActions;
};

export const bindActionsR = actions => {
  const wrappedActions = wrapActionsR(actions);

  return trigger => { 
    const actionAimer = createAction(trigger);

    return target => wrappedActions(actionAimer(target));
  };
};

export const createActionsR = actions => {
  const withAction = bindActionsR(actions);

  return trigger => {
    const withTrigger = withAction(trigger);

    return target => withTrigger(target);
  };
};
