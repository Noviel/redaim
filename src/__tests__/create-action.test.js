import { createAction } from '../actions';

describe('createAction', () => {  
  
  it('should return function when called once', () => {
    expect(typeof createAction()).toMatchSnapshot();
  });

  it('should return function when called twice', () => {
    expect(typeof createAction()()).toMatchSnapshot();
  });  
  
  it('should return object when called thrice', () => {
    expect(typeof createAction()()()).toMatchSnapshot();
  });

  it('should set trigger to default value if not specified', () => {
    const { type } = createAction()()();
    
    expect(type).toMatchSnapshot();
  });

  it('should set target to default value if not specified', () => {
    const { meta: { target } } = createAction()()();
    
    expect(target).toMatchSnapshot();
  });

  it('should set exec to default value if not specified', () => {
    const { meta: { exec } } = createAction()()();
    
    expect(exec).toMatchSnapshot();
  });

  it('should set action\'s type', () => {
    const v = createAction('OK_TYPE')()();

    expect(v.type).toMatchSnapshot();
  });

  it('should set action\'s meta.target', () => {
    const a = createAction('ACTION_TRIGGER')('OK_TARGET')();

    expect(a.meta.target).toMatchSnapshot();
  });

  it('should correctly merge action\'s properties', () => {
    const action = createAction('ACTION_TRIGGER')('target')({ 
      exec: 'EXEC_TYPE',
      payload: { data: 10 },
      meta: {
        obj: {
          inner: 10
        },
        value: 100
      }
    });

    expect(action).toMatchSnapshot();
  });

  it('should override action\'s type', () => {
    const action = createAction('OK')('target')({
      type: 'WRONG'
    });

    expect(action.type).toMatchSnapshot();
  });

  it('should override action\'s target', () => {
    const action = createAction('TYPE_FROM_TRIGGER')('OK_TARGET')({
      meta: { target: 'WRONG_TARGET' }
    });

    expect(action.meta.target).toMatchSnapshot();
  });

  it('should correctly merge custom action\'s properties', () => {
    const actionOne = createAction('MY_TRIGGER')('MY_TARGET')({
      exec: 'MY_EXEC',
      customProp: {
        value: 42
      }
    });

    const actionTwo = createAction('MY_TRIGGER')('MY_TARGET')({
      exec: 'MY_EXEC',
      payload: {
        data: 'payload-data'
      },
      meta: {
        val: 'meta-val'
      },
      customProp: {
        value: 42
      }
    });
    
    const toMatch = {
      actionOne,
      actionTwo
    };

    expect(toMatch).toMatchSnapshot();
  });

  it('should use exec from root', () => {
    const action = createAction('TRIGGER')('TARGET')({
      exec: 'EXEC_FROM_ROOT',
      payload: { data: 10 },
      meta: {
        exec: 'EXEC_FROM_META',
        value: 100
      }
    });

    expect(action.meta.exec).toMatchSnapshot();
  });

});
