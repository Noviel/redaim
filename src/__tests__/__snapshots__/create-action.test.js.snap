// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAction should correctly merge action's properties 1`] = `
Object {
  "meta": Object {
    "exec": "EXEC_TYPE",
    "obj": Object {
      "inner": 10,
    },
    "target": "target",
    "value": 100,
  },
  "payload": Object {
    "data": 10,
  },
  "type": "ACTION_TRIGGER",
}
`;

exports[`createAction should correctly merge custom action's properties 1`] = `
Object {
  "actionOne": Object {
    "customProp": Object {
      "value": 42,
    },
    "meta": Object {
      "exec": "MY_EXEC",
      "target": "MY_TARGET",
    },
    "payload": null,
    "type": "MY_TRIGGER",
  },
  "actionTwo": Object {
    "customProp": Object {
      "value": 42,
    },
    "meta": Object {
      "exec": "MY_EXEC",
      "target": "MY_TARGET",
      "val": "meta-val",
    },
    "payload": Object {
      "data": "payload-data",
    },
    "type": "MY_TRIGGER",
  },
}
`;

exports[`createAction should override action's target 1`] = `"OK_TARGET"`;

exports[`createAction should override action's type 1`] = `"OK"`;

exports[`createAction should return function when called once 1`] = `"function"`;

exports[`createAction should return function when called twice 1`] = `"function"`;

exports[`createAction should return object when called thrice 1`] = `"object"`;

exports[`createAction should set action's meta.target 1`] = `"OK_TARGET"`;

exports[`createAction should set action's type 1`] = `"OK_TYPE"`;

exports[`createAction should set exec to default value if not specified 1`] = `"@@REDAIM-ACTION-UNKNOWN"`;

exports[`createAction should set target to default value if not specified 1`] = `"@@REDAIM-TARGET-UNKNOWN"`;

exports[`createAction should set trigger to default value if not specified 1`] = `"@@REDAIM-TRIGGER-UNKNOWN"`;

exports[`createAction should use exec from root 1`] = `"EXEC_FROM_ROOT"`;
